### ISSUE TITLE
torch.multiprocessing.pool.Pool broken

### ISSUE THREAD

**Issue by lerks** [pytorch/pytorch#16954]
## 🐛 Bug

Cannot create a `torch.multiprocessing.pool.Pool` instance.

## To Reproduce

```
Python 3.6.3rc1+ (default, Feb  5 2019, 15:51:57)
Type 'copyright', 'credits' or 'license' for more information
IPython 6.5.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: from torch.multiprocessing.pool import Pool

In [2]: p = Pool(1)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-5-ce2ba693018c> in <module>()
----> 1 p = Pool(1)

/usr/local/fbcode/gcc-5-glibc-2.23/lib/python3.6/multiprocessing/pool.py in __init__(self, processes, initializer, initargs, maxtasksperchild, context)
    154                  maxtasksperchild=None, context=None):
    155         self._ctx = context or get_context()
--> 156         self._setup_queues()
    157         self._taskqueue = queue.Queue()
    158         self._cache = {}

/mnt/xarfuse/uid-169887/8eb75d00-ns-4026531840/torch/multiprocessing/pool.py in _setup_queues(self)
     21
     22     def _setup_queues(self):
---> 23         self._inqueue = SimpleQueue()
     24         self._outqueue = SimpleQueue()
     25         self._quick_put = self._inqueue._writer.send

TypeError: __init__() missing 1 required keyword-only argument: 'ctx'
```

## Expected behavior

No exception.

**Comment by lerks** [comment]
Sorry I didn't get back earlier on this. In the end I read the code further and stumbled on these lines:
https://github.com/pytorch/pytorch/blob/0fc03d155a91cbca2c9e23d23463f69fd254c0a3/torch/multiprocessing/__init__.py#L36-L40

They seem to hint to the fact that `torch.multiprocessing.pool.Pool` should in fact not be used directly, as it's only designed to be imported by PyTorch itself to replace the standard library's pool for versions of Python up to 3.2. Starting with Python 3.3, the standard library's pool will work without modifications.
These lines also hint that in fact the pool should be imported from `torch.multiprocessing`, not from `torch.multiprocessing.pool` (this is at odds with the documentation, which says that `torch.multiprocessing` exposes the same API as the standard `multiprocessing` package).

I've been been doing the above (in Python 3.6 and 3.7) and it all worked quite well. So, for me, this issue is now much less urgent than it originally was: it's just a matter of knowing where to import the pool from, and of perhaps updating the doc.

**Comment by lerks** [comment]
I think the issues in #17048 and #17155 are related. In some sense they stem from a problem that is already present in the standard library:
```
In [1]: from multiprocessing import Pool                                        

In [2]: type(Pool)                                                              
Out[2]: method

In [3]: from multiprocessing.pool import Pool                                   

In [4]: type(Pool)                                                              
Out[4]: type
```
You can see that while `multiprocessing.pool.Pool` is a class, `multiprocessing.pool` is a method of a context object, just like if you were doing `multiprocessing.get_context('spawn').Pool()` with an implicit context. Thus you can subclass the former, but not the latter.
PyTorch is exacerbating the issue because, due to the bug I reported here, the `torch.multiprocessing.pool.Pool` class doesn't work in Python 3.4+. A "solution" would be to first import `torch.multiprocessing` so that it can install the PyTorch reductions, but then subclass the standard library class `multiprocessing.pool.Pool`.

**Comment by igiloh** [comment]
@lerks - are you sure that:
```python
from torch import multiprocessing as mp
from multiprocessing.pool import Pool

class MyPool(Pool):
   pass
my_pool = MyPool()
```
Would end up with the exact same performance as `pool = mp.Pool()`? 

If I understand correctly - this essentially means that PyTorch reductions do get registered for `ForkingPickler` (that happens even just by `import torch` if I'm not mistaken) - but `_setup_queues()` doesn't get overridden to use `SimpleQueue()`s.  Won't that affect performance somehow?

**Comment by lerks** [comment]
From my current understanding of the code I think that, yes, these two approaches are equivalent. The reason I believe that is that in the code I linked above there's a comment saying "Later versions of Python already use ForkingPickler.", which I interpret as meaning "the SimpleQueue in Python 3.4+ gets automatic access to the reductions we install".

Though you may still want to wait for a PyTorch team member to weigh in and confirm.

**Comment by ailzhang** [comment]
I think it should be fine. cc: @ezyang  and @VitalyFedyunin might have better thoughts on this.

**Comment by ezyang** [comment]
> I've been been doing the above (in Python 3.6 and 3.7) and it all worked quite well. So, for me, this issue is now much less urgent than it originally was: it's just a matter of knowing where to import the pool from, and of perhaps updating the doc.

It sounds to me that if Python is sufficiently new, we should just expose `multiprocessing` directly (so there's no need for you to modify the import yourself). So that seems like a very reasonable fix, happy to review a patch to this effect.

**Comment by soumith** [comment]
@ezyang how would be override the pickler in such a case?

**Comment by ezyang** [comment]
Maybe I have misread the issue, but I read @lerks comment as implying that no override is necessary in recent versions of Python, as they default to ForkingPickler already.

**Comment by lerks** [comment]
Yes, that's my understanding from reading the code. It looks like importing `torch.multiprocessing` calls `init_reductions` which registers reducers on the standard library's `ForkingPickler`. So then using the standard `multiprocessing` module works just fine with tensors. I haven't double-checked that with a test though...

**Comment by Ge0rges** [comment]
So what would be the correct usage here? I see the issue s still open but it's been over a year so I'm wondering if this is still up for debate or there is accepted solution. Is it as simple as doing `from torch.multiprocessing import Pool` in python 3.7 and torch 1.3.1 for example?
